File: backend/src/guardian/Reporter.ts
/**
 * Reporter - 崩溃报告生成器
 * 为每次崩溃生成 Markdown 报告
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import { CrashInfo, AIDiagnosis, RepairAction, CrashReport } from './types.js';
import { logger } from '../utils/logger.js';

export interface ReportOptions {
  reportsDir: string;
}

export class Reporter {
  private readonly reportsDir: string;

  constructor(workDir: string) {
    this.reportsDir = path.join(workDir, '.guardian_reports');
    
    // 确保报告目录存在
    if (!fs.existsSync(this.reportsDir)) {
      fs.mkdirSync(this.reportsDir, { recursive: true });
    }
  }

  /**
   * 生成崩溃报告
   */
  public async generateReport(params: {
    serverDir: string;
    serverType: string;
    javaVersion: string;
    mcVersion: string;
    crashInfo: CrashInfo;
    diagnosis?: AIDiagnosis;
    executedActions: RepairAction[];
    result: 'fixed' | 'unfixed' | 'user_stopped' | 'give_up';
    restartCount: number;
  }): Promise<CrashReport> {
    const report: CrashReport = {
      id: `report_${Date.now()}_${Math.random().toString(36).substring(2, 6)}`,
      timestamp: new Date().toISOString(),
      serverDir: params.serverDir,
      serverType: params.serverType as any,
      javaVersion: params.javaVersion,
      mcVersion: params.mcVersion,
      crashInfo: params.crashInfo,
      diagnosis: params.diagnosis,
      executedActions: params.executedActions,
      result: params.result,
      restartCount: params.restartCount
    };

    // 生成 Markdown 内容
    const markdown = this.buildMarkdown(report);
    
    // 保存到文件
    const fileName = `crash_report_${report.id}.md`;
    const filePath = path.join(this.reportsDir, fileName);
    
    await fs.promises.writeFile(filePath, markdown, 'utf-8');
    
    report.reportPath = filePath;
    logger.info(`崩溃报告已保存: ${filePath}`);

    return report;
  }

  /**
   * 构建 Markdown 报告内容
   */
  private buildMarkdown(report: CrashReport): string {
    const { crashInfo, diagnosis, executedActions } = report;
    
    let md = `# 💥 服务端崩溃报告\n\n`;
    md += `> **报告ID**: ${report.id}\n`;
    md += `> **时间**: ${new Date(report.timestamp).toLocaleString('zh-CN')}\n`;
    md += `> **结果**: ${this.formatResult(report.result)}\n\n`;

    md += `---\n\n`;

    // 服务端信息
    md += `## 📋 服务端信息\n\n`;
    md += `| 项目 | 值 |\n`;
    md += `|------|-----|\n`;
    md += `| 服务端类型 | ${report.serverType} |\n`;
    md += `| Minecraft 版本 | ${report.mcVersion} |\n`;
    md += `| Java 版本 | ${report.javaVersion} |\n`;
    md += `| 工作目录 | \`${report.serverDir}\` |\n`;
    md += `| 重启次数 | ${report.restartCount} |\n\n`;

    // 崩溃信息
    md += `## 🚨 崩溃信息\n\n`;
    md += `- **严重等级**: ${this.formatSeverity(crashInfo.severity)}\n`;
    md += `- **退出码**: ${crashInfo.exitCode ?? 'N/A'}\n`;
    md += `- **信号**: ${crashInfo.signal ?? 'N/A'}\n`;
    md += `- **检测模式**: ${crashInfo.detectedPatterns.join(', ') || '无'}\n`;
    md += `- **分类**: ${crashInfo.classification.type}\n\n`;

    // 崩溃原因
    md += `### 崩溃原因\n\n`;
    md += `${crashInfo.classification.reason}\n\n`;

    // 疑似问题
    if (crashInfo.classification.suspectedMods.length > 0) {
      md += `### 疑似问题模组\n\n`;
      for (const mod of crashInfo.classification.suspectedMods) {
        md += `- \`${mod}\`\n`;
      }
      md += `\n`;
    }

    if (crashInfo.classification.suspectedConfigs.length > 0) {
      md += `### 疑似问题配置\n\n`;
      for (const cfg of crashInfo.classification.suspectedConfigs) {
        md += `- \`${cfg}\`\n`;
      }
      md += `\n`;
    }

    // AI 诊断
    if (diagnosis) {
      md += `## 🤖 AI 诊断\n\n`;
      md += `> ${diagnosis.diagnosis}\n\n`;
      
      md += `### 原因分析\n\n`;
      for (const cause of diagnosis.causes) {
        md += `1. ${cause}\n`;
      }
      md += `\n`;
      
      md += `**置信度**: ${Math.round(diagnosis.confidence * 100)}%\n\n`;
    }

    // 执行的操作
    if (executedActions.length > 0) {
      md += `## 🔧 已执行操作\n\n`;
      md += `| 操作 | 目标 | 结果 | 原因 |\n`;
      md += `|------|------|------|------|\n`;
      
      for (const action of executedActions) {
        md += `| ${action.type} | \`${action.target}\` | ${action.approved ? '✅ 已执行' : '⏳ 待确认'} | ${action.reason} |\n`;
      }
      md += `\n`;
    }

    // 日志上下文
    md += `## 📄 日志上下文（最后 ${crashInfo.logContext.length} 行）\n\n`;
    md += `\`\`\`log\n`;
    md += crashInfo.logContext.slice(-30).join('\n');
    md += `\n\`\`\`\n\n`;

    // 页脚
    md += `---\n`;
    md += `*由 ServerGuardian 自动生成*\n`;

    return md;
  }

  /**
   * 格式化结果
   */
  private formatResult(result: string): string {
    const map: Record<string, string> = {
      'fixed': '✅ 已修复',
      'unfixed': '❌ 未修复',
      'user_stopped': '🛑 用户终止',
      'give_up': '⚠️ 已放弃（多次崩溃）'
    };
    return map[result] || result;
  }

  /**
   * 格式化严重等级
   */
  private formatSeverity(severity: string): string {
    const map: Record<string, string> = {
      'fatal': '🔴 致命',
      'error': '🟠 错误',
      'warning': '🟡 警告',
      'info': '🔵 信息'
    };
    return map[severity] || severity;
  }

  /**
   * 获取所有报告列表
   */
  public getReportsList(): Array<{ id: string; timestamp: string; file: string }> {
    try {
      if (!fs.existsSync(this.reportsDir)) return [];

      const files = fs.readdirSync(this.reportsDir);
      const reports: Array<{ id: string; timestamp: string; file: string }> = [];

      for (const file of files) {
        if (!file.endsWith('.md')) continue;
        
        const match = file.match(/crash_report_(.+?)\.md/);
        if (match) {
          const filePath = path.join(this.reportsDir, file);
          const stat = fs.statSync(filePath);
          reports.push({
            id: match[1],
            timestamp: stat.mtime.toISOString(),
            file: filePath
          });
        }
      }

      return reports.sort((a, b) => new Date(b.timestamp).getTime() - new Date(a.timestamp).getTime());
    } catch (err) {
      logger.error('获取报告列表失败', err as Error);
      return [];
    }
  }

  /**
   * 获取报告内容
   */
  public getReportContent(reportId: string): string | null {
    try {
      const files = fs.readdirSync(this.reportsDir);
      for (const file of files) {
        if (file.includes(reportId)) {
          return fs.readFileSync(path.join(this.reportsDir, file), 'utf-8');
        }
      }
      return null;
    } catch {
      return null;
    }
  }
}

File: backend-net/src/DeEarthX.Guardian/Reporter.cs
using System.Text;
using System.Text.Json;
using DeEarthX.Core;
using DeEarthX.Core.Abstractions;

namespace DeEarthX.Guardian;

public sealed class Reporter
{
    private readonly IAppDirectoryProvider _appDir;
    private readonly ILogService _log;
    private string _reportsDir = string.Empty;

    public Reporter(IAppDirectoryProvider appDir, ILogService log)
    {
        _appDir = appDir;
        _log = log;
        _reportsDir = Path.Combine(appDir.GetAppDirectory(), "guardian", "reports");
        Directory.CreateDirectory(_reportsDir);
    }

    public void Configure(string _)
    {
        _reportsDir = Path.Combine(_appDir.GetAppDirectory(), "guardian", "reports");
        Directory.CreateDirectory(_reportsDir);
    }

    public async Task<CrashReport> GenerateReportAsync(ReportParams parameters, CancellationToken ct = default)
    {
        var report = new CrashReport
        {
            Id = $"report_{DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()}_{Guid.NewGuid():N}".Substring(0, 26),
            Timestamp = DateTime.UtcNow.ToString("o"),
            ServerDir = parameters.ServerDir,
            ServerType = parameters.ServerType,
            JavaVersion = parameters.JavaVersion,
            McVersion = parameters.McVersion,
            CrashInfo = parameters.CrashInfo,
            Diagnosis = parameters.Diagnosis,
            ExecutedActions = parameters.ExecutedActions,
            Result = parameters.Result,
            RestartCount = parameters.RestartCount
        };

        var markdown = BuildMarkdown(report);
        var fileName = $"crash_report_{report.Id}.md";
        var filePath = Path.Combine(_reportsDir, fileName);
        await File.WriteAllTextAsync(filePath, markdown, ct);
        report.ReportPath = filePath;
        _log.Info($"崩溃报告已保存: {filePath}");
        return report;
    }

    public List<ReportListItem> GetReportsList()
    {
        var reports = new List<ReportListItem>();
        if (!Directory.Exists(_reportsDir)) return reports;
        try
        {
            foreach (var file in Directory.EnumerateFiles(_reportsDir, "crash_report_*.md"))
            {
                var info = new FileInfo(file);
                var match = System.Text.RegularExpressions.Regex.Match(info.Name, @"crash_report_(.+)\.md");
                reports.Add(new ReportListItem
                {
                    Id = match.Success ? match.Groups[1].Value : info.Name,
                    Timestamp = info.LastWriteTimeUtc.ToString("o"),
                    File = file
                });
            }
        }
        catch (Exception ex)
        {
            _log.Error("获取报告列表失败", ex);
        }
        return reports.OrderByDescending(r => r.Timestamp).ToList();
    }

    public string? GetReportContent(string reportId)
    {
        if (!Directory.Exists(_reportsDir)) return null;
        try
        {
            foreach (var file in Directory.EnumerateFiles(_reportsDir, "*.md"))
            {
                if (Path.GetFileName(file).Contains(reportId, StringComparison.OrdinalIgnoreCase))
                {
                    return File.ReadAllText(file);
                }
            }
        }
        catch
        {
        }
        return null;
    }

    private static string BuildMarkdown(CrashReport report)
    {
        var crash = report.CrashInfo;
        var sb = new StringBuilder();
        sb.AppendLine("# 服务端崩溃报告");
        sb.AppendLine();
        sb.AppendLine($"> **报告ID**: {report.Id}");
        sb.AppendLine($"> **时间**: {report.Timestamp}");
        sb.AppendLine($"> **结果**: {FormatResult(report.Result)}");
        sb.AppendLine();
        sb.AppendLine("---");
        sb.AppendLine();
        sb.AppendLine("## 服务端信息");
        sb.AppendLine();
        sb.AppendLine("| 项目 | 值 |");
        sb.AppendLine("|------|-----|");
        sb.AppendLine($"| 服务端类型 | {report.ServerType} |");
        sb.AppendLine($"| Minecraft 版本 | {report.McVersion} |");
        sb.AppendLine($"| Java 版本 | {report.JavaVersion} |");
        sb.AppendLine($"| 工作目录 | `{report.ServerDir}` |");
        sb.AppendLine($"| 重启次数 | {report.RestartCount} |");
        sb.AppendLine();

        if (crash is not null)
        {
            sb.AppendLine("## 崩溃信息");
            sb.AppendLine();
            sb.AppendLine($"- **严重等级**: {FormatSeverity(crash.Severity)}");
            sb.AppendLine($"- **退出码**: {crash.ExitCode?.ToString() ?? "N/A"}");
            sb.AppendLine($"- **信号**: {crash.Signal ?? "N/A"}");
            sb.AppendLine($"- **检测模式**: {(crash.DetectedPatterns.Count > 0 ? string.Join(", ", crash.DetectedPatterns) : "无")}");
            sb.AppendLine($"- **分类**: {crash.Classification.Type}");
            sb.AppendLine();
            sb.AppendLine("### 崩溃原因");
            sb.AppendLine();
            sb.AppendLine(crash.Classification.Reason);
            sb.AppendLine();

            if (crash.Classification.SuspectedMods.Count > 0)
            {
                sb.AppendLine("### 疑似问题模组");
                sb.AppendLine();
                foreach (var mod in crash.Classification.SuspectedMods)
                {
                    sb.AppendLine($"- `{mod}`");
                }
                sb.AppendLine();
            }

            if (crash.Classification.SuspectedConfigs.Count > 0)
            {
                sb.AppendLine("### 疑似问题配置");
                sb.AppendLine();
                foreach (var cfg in crash.Classification.SuspectedConfigs)
                {
                    sb.AppendLine($"- `{cfg}`");
                }
                sb.AppendLine();
            }
        }

        if (report.Diagnosis is not null)
        {
            sb.AppendLine("## AI 诊断");
            sb.AppendLine();
            sb.AppendLine($"> {report.Diagnosis.Diagnosis}");
            sb.AppendLine();
            sb.AppendLine("### 原因分析");
            sb.AppendLine();
            for (var i = 0; i < report.Diagnosis.Causes.Count; i++)
            {
                sb.AppendLine($"{i + 1}. {report.Diagnosis.Causes[i]}");
            }
            sb.AppendLine();
            sb.AppendLine($"**置信度**: {Math.Round(report.Diagnosis.Confidence * 100)}%");
            sb.AppendLine();
        }

        if (report.ExecutedActions.Count > 0)
        {
            sb.AppendLine("## 已执行操作");
            sb.AppendLine();
            sb.AppendLine("| 操作 | 目标 | 结果 | 原因 |");
            sb.AppendLine("|------|------|------|------|");
            foreach (var action in report.ExecutedActions)
            {
                sb.AppendLine($"| {action.Type} | `{action.Target}` | {(action.Approved ? "已执行" : "待确认")} | {action.Reason} |");
            }
            sb.AppendLine();
        }

        if (crash is not null && crash.LogContext.Count > 0)
        {
            sb.AppendLine($"## 日志上下文（最后 {Math.Min(30, crash.LogContext.Count)} 行）");
            sb.AppendLine();
            sb.AppendLine("```log");
            var take = crash.LogContext.Count > 30 ? crash.LogContext.GetRange(crash.LogContext.Count - 30, 30) : crash.LogContext;
            sb.AppendLine(string.Join('\n', take));
            sb.AppendLine("```");
            sb.AppendLine();
        }

        sb.AppendLine("---");
        sb.AppendLine("*由 ServerGuardian 自动生成*");
        return sb.ToString();
    }

    private static string FormatResult(CrashReportResult result) => result switch
    {
        CrashReportResult.Fixed => "已修复",
        CrashReportResult.Unfixed => "未修复",
        CrashReportResult.UserStopped => "用户终止",
        CrashReportResult.GiveUp => "已放弃（多次崩溃）",
        _ => result.ToString()
    };

    private static string FormatSeverity(CrashSeverity severity) => severity switch
    {
        CrashSeverity.Fatal => "致命",
        CrashSeverity.Error => "错误",
        CrashSeverity.Warning => "警告",
        _ => "信息"
    };
}

public sealed class ReportParams
{
    public string ServerDir { get; set; } = string.Empty;
    public string ServerType { get; set; } = "unknown";
    public string JavaVersion { get; set; } = string.Empty;
    public string McVersion { get; set; } = string.Empty;
    public CrashInfo CrashInfo { get; set; } = new();
    public AiDiagnosisResult? Diagnosis { get; set; }
    public List<AiAction> ExecutedActions { get; set; } = new();
    public CrashReportResult Result { get; set; }
    public int RestartCount { get; set; }
}
